### PR TITLE
Feature/add addaptivecpp

### PR DESCRIPTION
**Pull request by will-saunders-ukaea**
Adds an AdaptiveCpp package.
I have left the OpenCl options in for now (default off). I think the configuration here is correct but I am yet to actually get acpp to work with an opencl implementation.

### CODE REVIEW

File: packages/adaptivecpp/package.py [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+                    "concretized llvm dependency must provide a "
+                    "unique directory containing libc++abi.so, "
+                    "found: {0}".format(so_paths)
+                )
+            rpaths.add(path.dirname(so_paths[0]))
+            default_cuda_link_line = "default-cuda-link-line"
+            if default_cuda_link_line in config.keys():
+                config[default_cuda_link_line] += " " + " ".join(
+                    "-rpath {0}".format(p) for p in rpaths
+                )
```
Comment (cmacmackin) [inline packages/adaptivecpp/package.py:248]: Should you also deal with the case where this key is not already present?

File: packages/adaptivecpp/package.py [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+        "boost@1.60.0: +filesystem +fiber +context cxxstd=17", when="@23.10.0:"
+    )
+    depends_on("python@3:")
+    # depends_on("llvm@8: +clang", when="~cuda")
+    depends_on("llvm@9: +clang", when="+cuda")
+    depends_on("llvm@9: +clang", when="+omp_llvm")
+    depends_on("cuda", when="@23.10.0: +cuda")
+    depends_on("cuda", when="+nvcxx")
+    depends_on("nvhpc@22.9:", when="+nvcxx", type="build")
+    depends_on("opencl@3.0", when="+opencl")
```
Comment (cmacmackin) [inline packages/adaptivecpp/package.py:78]: How does this work with different compilers? E.g., is it able to use GCC with `+nvcxx` or `+omp_llvm` variants? Or can it only use `nvcxx` and `clang` (respectively) in those cases?

**Review by cmacmackin — COMMENTED** [review]
This looks excellent. I just have a couple of questions before merging.

File: packages/adaptivecpp/package.py (hunk 1 above)
Comment (will-saunders-ukaea) [reply packages/adaptivecpp/package.py:248]: I think if they key is not present then nothing needs to be done. The hipsycl package (which was the source) assumes that the key is always present and this turned out not to be true in later versions.

File: packages/adaptivecpp/package.py (hunk 2 above)
Comment (will-saunders-ukaea) [reply packages/adaptivecpp/package.py:78]: I've almost exclusively used +nvcxx with GCC. I find that the nvhpc packages does not need to be loaded downstream when you use acpp to compile NESO (with cuda-nvcxx backend). But if you do set nvhpc to be a run dependency then the nvhpc package gets loaded when acpp gets loaded. Loading nvhpc then sets CC,CXX to be the nvidia compiler globally which breaks other compilation paths, e.g. omp.library-only.

Comment (will-saunders-ukaea) [reply packages/adaptivecpp/package.py:78]: I have not used omp_llvm for a while (again inherited from the hipsycl package) usually the host `omp.library-only` and `cuda-nvcxx` backends coexist within an single install without issue.

File: packages/adaptivecpp/package.py (hunk 1 above)
Comment (cmacmackin) [reply packages/adaptivecpp/package.py:248]: Well, if everything works then I guess it's fine as it is.

File: packages/adaptivecpp/package.py (hunk 2 above)
Comment (cmacmackin) [reply packages/adaptivecpp/package.py:78]: OK, I just wasn't sure how it all worked.

**Comment by cmacmackin** [comment]
To clarify, is this what the new releases of hipSYCL are called? Or is it something else? I keep losing track of the name changes.

**Comment by will-saunders-ukaea** [comment]
Correct - this is the continuation of hipsycl under the name AdaptiveCpp.

**Pull request merged**